File: modulo2/shopapp/app/src/main/java/com/shopapp/presentation/ui/uipublic/catalog/CatalogScreen.kt
// presentation/ui/uipublic/catalog/CatalogScreen.kt
package com.shopapp.presentation.ui.uipublic.catalog

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.*
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import com.shopapp.presentation.ui.uipublic.home.ProductCard
import com.shopapp.presentation.viewmodel.CatalogViewModel
import com.shopapp.theme.*

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CatalogScreen(
    onProductClick: (Int) -> Unit,
    viewModel: CatalogViewModel = hiltViewModel(),
) {
    val state    by viewModel.state.collectAsState()
    val gridState = rememberLazyGridState()

    // Detectar cuando llegar al final para cargar más
    val shouldLoadMore by remember {
        derivedStateOf {
            val lastVisible = gridState.layoutInfo.visibleItemsInfo.lastOrNull()?.index ?: 0
            val total       = gridState.layoutInfo.totalItemsCount
            lastVisible >= total - 3 && !state.isLoadingMore && state.hasMore
        }
    }

    LaunchedEffect(shouldLoadMore) {
        if (shouldLoadMore) viewModel.loadMore()
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Background),
    ) {
        // ── Barra de búsqueda ──────────────────────────────────
        Surface(color = Surface, tonalElevation = 0.dp) {
            Column(modifier = Modifier.padding(16.dp)) {

                // Título y contador
                Row(
                    modifier              = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment     = Alignment.CenterVertically,
                ) {
                    Text(
                        text       = "Catálogo",
                        style      = MaterialTheme.typography.headlineMedium,
                        fontWeight = FontWeight.Bold,
                        color      = TextPrimary,
                    )
                    Text(
                        text  = "${state.total} productos",
                        style = MaterialTheme.typography.bodySmall,
                        color = TextSecondary,
                    )
                }

                Spacer(Modifier.height(12.dp))

                // Campo de búsqueda
                OutlinedTextField(
                    value         = state.search,
                    onValueChange = viewModel::setSearch,
                    placeholder   = { Text("Buscar productos...", color = TextFaint) },
                    leadingIcon   = { Icon(Icons.Default.Search, null, tint = TextSecondary) },
                    singleLine    = true,
                    modifier      = Modifier.fillMaxWidth(),
                    shape         = MaterialTheme.shapes.medium,
                    colors        = OutlinedTextFieldDefaults.colors(
                        focusedBorderColor   = Accent,
                        unfocusedBorderColor = Border,
                        cursorColor          = Accent,
                    ),
                )

                Spacer(Modifier.height(12.dp))

                // Chips de ordenamiento
                LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    items(listOf(
                        "" to "Relevancia",
                        "price"  to "Precio ↑",
                        "-price" to "Precio ↓",
                        "-created_at" to "Recientes",
                    )) { (value, label) ->
                        FilterChip(
                            selected = state.ordering == value,
                            onClick  = { viewModel.setOrdering(value) },
                            label    = { Text(label, style = MaterialTheme.typography.labelSmall) },
                            colors   = FilterChipDefaults.filterChipColors(
                                selectedContainerColor    = Accent,
                                selectedLabelColor        = AccentOnDark,
                                containerColor            = Surface2,
                                labelColor                = TextSecondary,
                            ),
                        )
                    }
                }

                // Chips de categorías
                if (state.categories.isNotEmpty()) {
                    Spacer(Modifier.height(8.dp))
                    LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        item {
                            FilterChip(
                                selected = state.selectedCategory == null,
                                onClick  = { viewModel.setCategory(null) },
                                label    = { Text("Todas", style = MaterialTheme.typography.labelSmall) },
                                colors   = FilterChipDefaults.filterChipColors(
                                    selectedContainerColor = Accent,
                                    selectedLabelColor     = AccentOnDark,
                                    containerColor         = Surface2,
                                    labelColor             = TextSecondary,
                                ),
                            )
                        }
                        items(state.categories) { cat ->
                            FilterChip(
                                selected = state.selectedCategory == cat.id,
                                onClick  = { viewModel.setCategory(cat.id) },
                                label    = { Text(cat.name, style = MaterialTheme.typography.labelSmall) },
                                colors   = FilterChipDefaults.filterChipColors(
                                    selectedContainerColor = Accent,
                                    selectedLabelColor     = AccentOnDark,
                                    containerColor         = Surface2,
                                    labelColor             = TextSecondary,
                                ),
                            )
                        }
                    }
                }
            }
        }

        // ── Grid de productos ──────────────────────────────────
        when {
            state.isLoading -> {
                Box(Modifier.fillMaxSize(), Alignment.Center) {
                    CircularProgressIndicator(color = Accent)
                }
            }
            state.error != null -> {
                Box(Modifier.fillMaxSize(), Alignment.Center) {
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Text("❌ ${state.error}", color = Error)
                        Spacer(Modifier.height(12.dp))
                        Button(
                            onClick = viewModel::refresh,
                            colors  = ButtonDefaults.buttonColors(containerColor = Accent),
                        ) { Text("Reintentar", color = AccentOnDark) }
                    }
                }
            }
            state.products.isEmpty() -> {
                Box(Modifier.fillMaxSize(), Alignment.Center) {
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Text("🔍", style = MaterialTheme.typography.displayMedium)
                        Spacer(Modifier.height(8.dp))
                        Text("Sin resultados", color = TextPrimary, fontWeight = FontWeight.Bold)
                        Text("Intenta con otra búsqueda", color = TextSecondary)
                    }
                }
            }
            else -> {
                LazyVerticalGrid(
                    columns       = GridCells.Fixed(2),
                    state         = gridState,
                    contentPadding = PaddingValues(16.dp),
                    verticalArrangement   = Arrangement.spacedBy(12.dp),
                    horizontalArrangement = Arrangement.spacedBy(12.dp),
                    modifier      = Modifier.fillMaxSize(),
                ) {
                    items(state.products, key = { it.id }) { product ->
                        ProductCard(
                            product = product,
                            onClick = { onProductClick(product.id) },
                        )
                    }

                    if (state.isLoadingMore) {
                        item(span = { GridItemSpan(2) }) {
                            Box(
                                modifier         = Modifier.fillMaxWidth().padding(16.dp),
                                contentAlignment = Alignment.Center,
                            ) {
                                CircularProgressIndicator(
                                    color       = Accent,
                                    modifier    = Modifier.size(28.dp),
                                    strokeWidth = 2.dp,
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}

File: modulo4/flutter_shop_app/lib/presentation/screens/catalog/catalog_screen.dart
// lib/presentation/screens/catalog/catalog_screen.dart

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import '../../../theme/app_colors.dart';
import '../../providers/catalog_provider.dart';
import '../../widgets/product_card.dart';

class CatalogScreen extends ConsumerStatefulWidget {
  const CatalogScreen({super.key});

  @override
  ConsumerState<CatalogScreen> createState() => _CatalogScreenState();
}

class _CatalogScreenState extends ConsumerState<CatalogScreen> {
  final _searchCtrl  = TextEditingController();
  final _scrollCtrl  = ScrollController();

  @override
  void initState() {
    super.initState();
    _scrollCtrl.addListener(_onScroll);
  }

  @override
  void dispose() {
    _searchCtrl.dispose();
    _scrollCtrl.dispose();
    super.dispose();
  }

  void _onScroll() {
    if (_scrollCtrl.position.pixels >= _scrollCtrl.position.maxScrollExtent - 200) {
      ref.read(catalogProvider.notifier).loadMore();
    }
  }

  @override
  Widget build(BuildContext context) {
    final state        = ref.watch(catalogProvider);
    final catsAsync    = ref.watch(categoriesProvider);
    final tt           = Theme.of(context).textTheme;

    return Scaffold(
      body: SafeArea(
        child: Column(
          children: [
            // ── Header ──────────────────────────────────────
            Container(
              color:   AppColors.surface,
              padding: const EdgeInsets.fromLTRB(16, 16, 16, 0),
              child:   Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    children: [
                      Text('Catálogo', style: tt.headlineMedium),
                      Text(
                        '${state.total} productos',
                        style: tt.bodySmall,
                      ),
                    ],
                  ),
                  const SizedBox(height: 12),

                  // Búsqueda
                  TextField(
                    controller:  _searchCtrl,
                    onChanged:   (q) {
                      ref.read(catalogProvider.notifier).setSearch(q);
                    },
                    decoration: const InputDecoration(
                      hintText:    'Buscar productos...',
                      prefixIcon:  Icon(Icons.search_rounded, color: AppColors.textSecondary),
                      contentPadding: EdgeInsets.symmetric(vertical: 12),
                    ),
                    style: const TextStyle(color: AppColors.textPrimary),
                  ),
                  const SizedBox(height: 10),

                  // Chips de ordenamiento
                  SizedBox(
                    height: 34,
                    child: ListView(
                      scrollDirection: Axis.horizontal,
                      children: [
                        for (final item in [
                          ('',             'Relevancia'),
                          ('price',        'Precio ↑'),
                          ('-price',       'Precio ↓'),
                          ('-created_at',  'Recientes'),
                        ])
                          Padding(
                            padding: const EdgeInsets.only(right: 8),
                            child:   ChoiceChip(
                              label:     Text(item.$2),
                              selected:  state.ordering == item.$1,
                              onSelected:(_) => ref.read(catalogProvider.notifier).setOrdering(item.$1),
                            ),
                          ),
                      ],
                    ),
                  ),
                  const SizedBox(height: 8),

                  // Chips de categorías
                  catsAsync.when(
                    loading: () => const SizedBox.shrink(),
                    error:   (_, __) => const SizedBox.shrink(),
                    data: (cats) => SizedBox(
                      height: 34,
                      child: ListView(
                        scrollDirection: Axis.horizontal,
                        children: [
                          Padding(
                            padding: const EdgeInsets.only(right: 8),
                            child:   ChoiceChip(
                              label:     const Text('Todas'),
                              selected:  state.selectedCategory == null,
                              onSelected:(_) => ref.read(catalogProvider.notifier).setCategory(null),
                            ),
                          ),
                          for (final cat in cats.where((c) => c.isActive))
                            Padding(
                              padding: const EdgeInsets.only(right: 8),
                              child:   ChoiceChip(
                                label:     Text(cat.name),
                                selected:  state.selectedCategory == cat.id,
                                onSelected:(_) =>
                                    ref.read(catalogProvider.notifier).setCategory(cat.id),
                              ),
                            ),
                        ],
                      ),
                    ),
                  ),
                  const SizedBox(height: 12),
                ],
              ),
            ),

            // ── Grid de productos ────────────────────────────
            Expanded(
              child: Builder(
                builder: (_) {
                  if (state.isLoading && state.products.isEmpty) {
                    return const Center(
                      child: CircularProgressIndicator(color: AppColors.accent),
                    );
                  }
                  if (state.error != null && state.products.isEmpty) {
                    return Center(
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          const Text('❌', style: TextStyle(fontSize: 40)),
                          const SizedBox(height: 12),
                          Text(state.error!, style: const TextStyle(color: AppColors.error)),
                          const SizedBox(height: 16),
                          FilledButton(
                            onPressed: () => ref.read(catalogProvider.notifier).refresh(),
                            style:     FilledButton.styleFrom(
                              backgroundColor: AppColors.accent,
                              foregroundColor: AppColors.onAccent,
                            ),
                            child: const Text('Reintentar'),
                          ),
                        ],
                      ),
                    );
                  }
                  if (state.products.isEmpty) {
                    return const Center(
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          Text('🔍', style: TextStyle(fontSize: 48)),
                          SizedBox(height: 12),
                          Text('Sin resultados', style: TextStyle(color: AppColors.textPrimary, fontSize: 18, fontWeight: FontWeight.bold)),
                          Text('Intenta con otra búsqueda', style: TextStyle(color: AppColors.textSecondary)),
                        ],
                      ),
                    );
                  }

                  return GridView.builder(
                    controller:  _scrollCtrl,
                    padding:     const EdgeInsets.all(16),
                    gridDelegate:const SliverGridDelegateWithFixedCrossAxisCount(
                      crossAxisCount:   2,
                      crossAxisSpacing: 12,
                      mainAxisSpacing:  12,
                      childAspectRatio: 0.68,
                    ),
                    itemCount:  state.products.length + (state.isLoadingMore ? 1 : 0),
                    itemBuilder:(ctx, i) {
                      if (i >= state.products.length) {
                        return const Center(
                          child: Padding(
                            padding: EdgeInsets.all(16),
                            child:   CircularProgressIndicator(
                              color:       AppColors.accent,
                              strokeWidth: 2,
                            ),
                          ),
                        );
                      }
                      final product = state.products[i];
                      return ProductCard(
                        product: product,
                        onTap:   () => context.push('/catalog/${product.id}'),
                      );
                    },
                  );
                },
              ),
            ),
          ],
        ),
      ),
    );
  }
}